File: tools/jsonld-js/verify_openidentity_context.mjs
#!/usr/bin/env node
/**
 * Independent jsonld.js verifier for OpenIdentity JSON-LD context candidate.
 *
 * Usage:
 *   npm install
 *   node verify_openidentity_context.mjs ../../contexts/openidentity-v1.jsonld
 */
import fs from "node:fs/promises";
import process from "node:process";
import jsonld from "jsonld";

const NS = "https://openidentity.foundation/ns#";
const XSD_INT = "http://www.w3.org/2001/XMLSchema#integer";
const URL = "https://openidentity.foundation/ns/v1";

function fail(m) {
    console.error("\nFAILED\n------\n" + m + "\n");
    process.exit(1);
}

function ok(m) {
    console.log("  " + m + ": PASS");
}

const file = process.argv[2];
if (!file) fail("Usage: node verify_openidentity_context.mjs <context-file>");

let ctx;
try {
    ctx = JSON.parse(await fs.readFile(file, "utf8"));
} catch (e) {
    fail("Unable to load context: " + e.message);
}

const documentLoader = async url => {
    if (url === URL) return {contextUrl: null, documentUrl: url, document: ctx};
    throw new Error("Remote context blocked by verifier: " + url);
};

console.log("\nOpenIdentity jsonld.js Context Verification\n" + "=".repeat(48));

const c = ctx["@context"];
const required = ["openIdentityVerificationPolicy", "appliesTo", "threshold", "Single", "Threshold"];
if (!c || typeof c !== "object" || c["@protected"] !== true || required.some(x => !(x in c)))
    fail("J01 context structure invalid");
ok("J01 context structure/load");

const local = {
    id: "@id", type: "@type",
    verificationMethod: {
        "@id": "https://www.w3.org/ns/did#verificationMethod",
        "@type": "@id", "@container": "@set"
    },
    capabilityInvocation: {
        "@id": "https://www.w3.org/ns/did#capabilityInvocation",
        "@type": "@id", "@container": "@set"
    }
};
const tc = [URL, local];

const thresholdDoc = {
    "@context": tc,
    id: "did:open:zExample#controller-policy",
    type: "Threshold",
    appliesTo: "capabilityInvocation",
    threshold: 2,
    verificationMethod: ["did:open:zExample#vm-A", "did:open:zExample#vm-B"]
};

let expanded;
try {
    expanded = await jsonld.expand(thresholdDoc, {documentLoader});
} catch (e) {
    fail("J02 expansion failed: " + e.message);
}
if (!expanded.length || !(expanded[0]["@type"] || []).includes(NS + "Threshold"))
    fail("J02 Threshold type expansion incorrect");
const tv = expanded[0][NS + "threshold"];
if (!tv?.length || tv[0]["@value"] !== 2 || tv[0]["@type"] !== XSD_INT)
    fail("J02 threshold datatype/value incorrect");
ok("J02 Threshold expansion");

const singleDoc = {...thresholdDoc, type: "Single", verificationMethod: ["did:open:zExample#vm-A"]};
delete singleDoc.threshold;
let sx;
try {
    sx = await jsonld.expand(singleDoc, {documentLoader});
} catch (e) {
    fail("J03 SINGLE expansion failed: " + e.message);
}
if (!sx.length || !(sx[0]["@type"] || []).includes(NS + "Single"))
    fail("J03 Single type expansion incorrect");
ok("J03 SINGLE expansion");

let compact, reexpanded;
try {
    compact = await jsonld.compact(expanded, tc, {documentLoader});
    reexpanded = await jsonld.expand(compact, {documentLoader});
} catch (e) {
    fail("J04 round trip failed: " + e.message);
}
const canon = x => JSON.stringify(x, (k, v) => {
    if (Array.isArray(v)) return v.map(y => y);
    if (v && typeof v === "object") return Object.fromEntries(Object.entries(v).sort(([a], [b]) => a.localeCompare(b)));
    return v;
});
if (canon(expanded) !== canon(reexpanded)) fail("J04 semantic round trip changed expansion");
ok("J04 semantic round trip");

const bad = {"@context": [URL, {threshold: "https://evil.example/threshold"}], threshold: 2};
let rejected = false;
try {
    await jsonld.expand(bad, {documentLoader});
} catch {
    rejected = true;
}
if (!rejected) fail("J05 protected term redefinition accepted");
ok("J05 protected-term redefinition rejected");

if ("@vocab" in c) fail("J06 context MUST NOT define @vocab");
let ux;
try {
    ux = await jsonld.expand({"@context": URL, totallyUnknownOpenIdentityTerm: "value"}, {documentLoader});
} catch (e) {
    fail("J06 unknown-term expansion failed: " + e.message);
}
if (ux.length && (NS + "totallyUnknownOpenIdentityTerm") in ux[0])
    fail("J06 unknown term leaked into OpenIdentity namespace");
ok("J06 no accidental vocabulary leakage");

const vp = c.openIdentityVerificationPolicy;
if (vp && typeof vp === "object" && vp["@type"] === "@id")
    fail("J07 openIdentityVerificationPolicy incorrectly coerces embedded object to @id");
ok("J07 embedded policy property has no @id coercion");

console.log("\n" + "=".repeat(48));
console.log("OPENIDENTITY JSONLD.JS CONTEXT CANDIDATE PASSED");
console.log("=".repeat(48) + "\n");

File: tools/jsonld/verify_openidentity_context.py
import argparse, json, sys
from pathlib import Path

try:
    from pyld import jsonld
except ImportError:
    print("Install PyLD: pip install PyLD", file=sys.stderr);
    sys.exit(2)
NS = "https://openidentity.foundation/ns#";
XSD = "http://www.w3.org/2001/XMLSchema#integer";
URL = "https://openidentity.foundation/ns/v1"


def fail(m): print("\nFAILED\n------\n" + m);sys.exit(1)


def ok(m): print("  " + m + ": PASS")


def main():
    p = argparse.ArgumentParser();
    p.add_argument("context_file", type=Path);
    a = p.parse_args()
    ctx = json.loads(a.context_file.read_text(encoding="utf-8"));
    c = ctx.get("@context")
    if not isinstance(c, dict): fail("Missing object @context")

    def loader(url, options=None):
        if url == URL: return {"contextUrl": None, "documentUrl": url, "document": ctx}
        raise jsonld.JsonLdError("Remote context blocked", "loading document failed", details={"url": url})

    print("\nOpenIdentity JSON-LD Context Verification\n" + "=" * 48)
    req = {"openIdentityVerificationPolicy", "appliesTo", "threshold", "Single", "Threshold"}
    if c.get("@protected") is not True or not req.issubset(c): fail("C01 structure invalid")
    ok("C01 context structure/load")
    local = {"id": "@id", "type": "@type",
             "verificationMethod": {"@id": "https://www.w3.org/ns/did#verificationMethod", "@type": "@id",
                                    "@container": "@set"},
             "capabilityInvocation": {"@id": "https://www.w3.org/ns/did#capabilityInvocation", "@type": "@id",
                                      "@container": "@set"}}
    tc = [URL, local]
    d = {"@context": tc, "id": "did:open:zExample#controller-policy", "type": "Threshold",
         "appliesTo": "capabilityInvocation", "threshold": 2,
         "verificationMethod": ["did:open:zExample#vm-A", "did:open:zExample#vm-B"]}
    try:
        e = jsonld.expand(d, options={"documentLoader": loader})
    except Exception as x:
        fail("C02 expansion: " + str(x))
    if not e or NS + "Threshold" not in e[0].get("@type", []): fail("C02 Threshold type wrong")
    vals = e[0].get(NS + "threshold", [])
    if not vals or vals[0].get("@value") != 2 or vals[0].get("@type") != XSD: fail("C02 threshold wrong")
    ok("C02 Threshold expansion")
    s = dict(d);
    s["type"] = "Single";
    s.pop("threshold");
    s["verificationMethod"] = ["did:open:zExample#vm-A"]
    try:
        se = jsonld.expand(s, options={"documentLoader": loader})
    except Exception as x:
        fail("C03 expansion: " + str(x))
    if not se or NS + "Single" not in se[0].get("@type", []): fail("C03 Single type wrong")
    ok("C03 SINGLE expansion")
    try:
        compact = jsonld.compact(e, tc, options={"documentLoader": loader});
        e2 = jsonld.expand(compact, options={"documentLoader": loader})
    except Exception as x:
        fail("C04 round trip: " + str(x))
    if json.dumps(e, sort_keys=True) != json.dumps(e2, sort_keys=True): fail(
        "C04 semantic round trip changed expansion")
    ok("C04 semantic round trip")
    bad = {"@context": [URL, {"threshold": "https://evil.example/threshold"}], "threshold": 2}
    try:
        jsonld.expand(bad, options={"documentLoader": loader})
    except Exception:
        ok("C05 protected-term redefinition rejected")
    else:
        fail("C05 protected term redefinition accepted")
    if "@vocab" in c: fail("C06 context MUST NOT define @vocab")
    u = {"@context": URL, "totallyUnknownOpenIdentityTerm": "value"}
    try:
        ue = jsonld.expand(u, options={"documentLoader": loader})
    except Exception as x:
        fail("C06 expansion: " + str(x))
    if ue and NS + "totallyUnknownOpenIdentityTerm" in ue[0]: fail("C06 unknown term leaked into OI vocabulary")
    ok("C06 no accidental vocabulary leakage")
    vp = c["openIdentityVerificationPolicy"]
    if isinstance(vp, dict) and vp.get("@type") == "@id": fail(
        "DESIGN REVIEW: remove @type:@id from openIdentityVerificationPolicy; its value is an embedded policy object.")
    print("\n" + "=" * 48 + "\nOPENIDENTITY JSON-LD CONTEXT CANDIDATE PASSED\n" + "=" * 48)


if __name__ == "__main__": main()
